### PR TITLE
Fix Decimal Mode (#145)

### LINKED ISSUE #145: [BUG] BCD Addition not Propagating Carry
The following code should set the carry flag, but does not:

SED; LDA #$80; ADC #$80

### PR DESCRIPTION
**Pull request by qbradq**

### CODE REVIEW

**Comment by qbradq** [comment]
Please do not merge yet. On further analysis this only fixes this particular symptom of the incorrect BCD behavior. I am working on the complete patch.

File: cpu/instructions.h [hunk 1]
```diff
@@ -16,44 +16,53 @@
...
-    result = (uint16_t)a + value + (uint16_t)(status & FLAG_CARRY);
-   
-    carrycalc(result);
-    zerocalc(result);
-    overflowcalc(result, a, value);
-    signcalc(result);
-    
     #ifndef NES_CPU
     if (status & FLAG_DECIMAL) {
         clearcarry();
```
Comment (greg-king5) [inline cpu/instructions.h]: That line is a left-over from the bug -- it must be removed (previous carries must be included in additions).

(It's one of two reasons why your use-case still fails.)

File: cpu/instructions.h [hunk 2]
```diff
@@ -16,44 +16,53 @@
...
         }
-        if ((result & 0xF0) > 0x90) {
-            result += 0x60;
+        if (tmp2 > 0x90) {
+            tmp2 += 0x60;
+        }
+        if (tmp2 & 0xFF00) {
             setcarry();
         }
+        result = (tmp & 0x0F) + (tmp2 & 0xF0);
```
Comment (greg-king5) [inline cpu/instructions.h]: I would use the or operator (`'|'`) instead of addition.

**Comment by qbradq** [comment]
I am trying to find a 6502 test suite that I can figure out how to run so I can be reasonably sure of ADC/SBC behavior.

**Comment by Kroc** [comment]
6502/65C02 test (including decimal mode), in A65 format: https://github.com/Klaus2m5/6502_65C02_functional_tests

This page is useful for other test suites, including decimal-specific test: http://visual6502.org/wiki/index.php?title=6502TestPrograms

**Comment by qbradq** [comment]
Those are the test suites I have been trying to get running :) I took a crack at SBC anyway with accompanying screenshot.


I have been double-checking the implementations against VICE and MAME.

**Comment by mist64** [comment]
We need to be careful to compare this with the 65C02, not the 6502. They behave different in decimal mode. Does one of the test suites test the 65C02 in particular?

**Comment by mist64** [comment]
@qbradq I can see you closed one of the bugs as "fixed". But this pull request is not merged yet. Is this the correct version to merge?

(The bug will be closed automatically when the pull request (referencing it correctly) is merged.)

**Comment by qbradq** [comment]
> @qbradq I can see you closed one of the bugs as "fixed". But this pull request is not merged yet. Is this the correct version to merge?
> 
> (The bug will be closed automatically when the pull request (referencing it correctly) is merged.)

I guess I should have closed the other as duplicate. I just didn't want multiple issues referring to the same symptom.

**Comment by tonu1** [comment]
ef5be9da0fcfc8eb8a1e43686bcf61f2098a52ee

**Comment by MJoergen** [comment]
> 6502/65C02 test (including decimal mode), in A65 format: https://github.com/Klaus2m5/6502_65C02_functional_tests

For what it's worth, I've ported this test suite (to test my own 65c02 emulator) to ca65 format here: https://github.com/MJoergen/cpu65c02/blob/master/65C02_extended_opcodes_test.s
Note that I've purposely commented out the decimal mode test, because my emulator does NOT support decimal mode. But the test suite should run just fine.

The lines to uncomment are lines 2287-2389.

**Pull request merged**